### PR TITLE
feat: add support for CORS as a first class citizen 

### LINKED ISSUE #1019: Support CORS configuration as a first class citizen in buckets
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

 **Is your feature request related to a problem? Please describe.**

Yes, there was an internal YAQS request around confusion on how to use CORS since every other client library exposes it like a first class citizen. 

**A clear and concise description of what you want to happen.**

Support for CORS to be clear in the documentation and for anyone looking for it.

### PR DESCRIPTION
**Pull request by hilts-vaughan**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-storage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass; I ran NPM lint and a bunch of unrelated things came up
- [?] Code coverage does not decrease (if any source code was changed)
- [?] Appropriate docs were updated (if necessary)

Fixes #1019

### CODE REVIEW

**Comment by hilts-vaughan** [comment]
Do you want me to add a CORS sample or is what we have enough?

**Comment by hilts-vaughan** [comment]
Looks like some of the system tests need some quick changes. Any reco's there?

File: src/bucket.ts [hunk 1]
```diff
@@ -2874,6 +2875,47 @@ class Bucket extends ServiceObject {
...
+   *
+   * @param {Cors[]} corsConfiguration The new CORS configuration to set
+   * @param {SetBucketMetadataCallback} [callback] Callback function.
+   * @returns {Promise<SetBucketMetadataResponse>}
+   *
+   * @example
+   * const storage = require('@google-cloud/storage')();
+   * const bucket = storage.bucket('albums');
+   *
+   * const corsConfiguration = [{"maxAgeSeconds": 3600}]; // 1 hour
```
Comment (stephenplusplus) [inline src/bucket.ts]: You can lose the quotation marks around `maxAgeSeconds` :)

File: system-test/storage.ts [hunk 2]
```diff
@@ -1217,6 +1217,66 @@ describe('storage', () => {
...
 
+  describe('cors configuration', () => {
+    const corsEntry = [
+      {
+        maxAgeSeconds: 1600
+      },
+      {
+        maxAgeSeconds: 3600,
+        method: [
+          "GET",
```
Comment (stephenplusplus) [inline system-test/storage.ts]: Can you update all `"`s to `'`s? This will make it compliant with our linter.

File: system-test/storage.ts [hunk 3]
```diff
@@ -1217,6 +1217,66 @@ describe('storage', () => {
...
+        await storage.createBucket(bucket.name, {
+          cors: corsEntry,
+        });
+
+        await bucket.getMetadata();
+        assert.strictEqual(
+          bucket.metadata.cors,
+          corsEntry
+        );
+      })
```
Comment (stephenplusplus) [inline system-test/storage.ts]: Missing trailing `;`

File: system-test/storage.ts [hunk 4]
```diff
@@ -1217,6 +1217,66 @@ describe('storage', () => {
...
+
+    describe('bucket', () => {
+      it('should create a bucket with a CORS configuration when passed in', async () => {
+        const bucket = storage.bucket(generateName());
+        await storage.createBucket(bucket.name, {
+          cors: corsEntry,
+        });
+
+        await bucket.getMetadata();
+        assert.strictEqual(
```
Comment (AVaksman) [inline system-test/storage.ts]:
```suggestion
        assert.deepStrictEqual(
```

File: system-test/storage.ts [hunk 5]
```diff
@@ -1217,6 +1217,66 @@ describe('storage', () => {
...
+          corsEntry
+        );
+      })
+
+      it('should set a CORS configuration', async () => {
+        const bucket = storage.bucket(generateName());
+        await bucket.create();
+        await bucket.setCorsConfiguration(corsEntry);
+        await bucket.getMetadata();
+        assert.strictEqual(
```
Comment (AVaksman) [inline system-test/storage.ts]:
```suggestion
        assert.deepStrictEqual(
```

File: system-test/storage.ts [hunk 6]
```diff
@@ -1217,6 +1217,66 @@ describe('storage', () => {
...
+          bucket.metadata.cors,
+          corsEntry
+        );
+        
+        // And now test the removing
+        bucket.setCorsConfiguration([]);
```
Comment (AVaksman) [inline system-test/storage.ts]:
```suggestion
        await bucket.setCorsConfiguration([]);
```

File: system-test/storage.ts [hunk 7]
```diff
@@ -1217,6 +1217,66 @@ describe('storage', () => {
...
+        assert.strictEqual(bucket.metadata.cors.length, 0);
```
Comment (AVaksman) [inline system-test/storage.ts]: after removing the CORS configs `bucket.metadata.cors` will be `undefined`
```suggestion
        assert.ok(!bucket.metadata.cors);
```

**Comment by AVaksman** [comment]
> Looks like some of the system tests need some quick changes. Any reco's there?

@hilts-vaughan thank you very much for the contribution.
Provided recommendations for the system tests.

running `npm run fix` should make the linter happy.

File: src/bucket.ts [hunk 8]
```diff
@@ -2825,6 +2825,7 @@ class Bucket extends ServiceObject {
     duration: number,
     callback: SetBucketMetadataCallback
   ): void;
+
```
Comment (AVaksman) [inline src/bucket.ts]: nit: unnecessary empty line

File: src/bucket.ts [hunk 9]
```diff
@@ -2874,6 +2875,47 @@ class Bucket extends ServiceObject {
     );
   }
 
+
```
Comment (AVaksman) [inline src/bucket.ts]: nit: unnecessary empty line

File: src/bucket.ts (hunk 1 above)
Comment (hilts-vaughan) [reply src/bucket.ts]: :+1:

File: system-test/storage.ts (hunk 2 above)
Comment (hilts-vaughan) [reply system-test/storage.ts]: Did a linter fix.

File: src/bucket.ts (hunk 8 above)
Comment (hilts-vaughan) [reply src/bucket.ts]: Resolved.

File: src/bucket.ts (hunk 9 above)
Comment (hilts-vaughan) [reply src/bucket.ts]: Resolved.

File: system-test/storage.ts (hunk 4 above)
Comment (hilts-vaughan) [reply system-test/storage.ts]: Done.

File: system-test/storage.ts (hunk 3 above)
Comment (hilts-vaughan) [reply system-test/storage.ts]: Should be good now.

File: system-test/storage.ts (hunk 5 above)
Comment (hilts-vaughan) [reply system-test/storage.ts]: Should be good now.

File: system-test/storage.ts (hunk 7 above)
Comment (hilts-vaughan) [reply system-test/storage.ts]: I see that now; thanks!

File: system-test/storage.ts (hunk 6 above)
Comment (hilts-vaughan) [reply system-test/storage.ts]: Thank you!

File: src/bucket.ts [hunk 10]
```diff
@@ -2873,6 +2873,47 @@ class Bucket extends ServiceObject {
...
+    corsConfiguration: Cors[]
+  ): Promise<SetBucketMetadataResponse>;
+  setCorsConfiguration(
+    corsConfiguration: Cors[],
+    callback: SetBucketMetadataCallback
+  ): void;
+  /**
+   * This can be used to set the CORS configuration on the bucket.
+   *
+   * The configuration will be overwriteten with the value passed into this.
```
Comment (stephenplusplus) [inline src/bucket.ts]: overwritten*

**Comment by hilts-vaughan** [comment]
Is that everything more or less?

**Comment by stephenplusplus** [comment]
Thanks!

**Pull request merged**